### PR TITLE
[ops, dist] feat: add FlexAttention with Ulysses sequence parallelism

### PR DESCRIPTION
**Pull request by Zx55**
### Why

Some Transformer models require two-dimensional attention masks, such as mixed full/sliding visibility. FlashAttention cannot represent these masks and must fall back to SDPA, whose dense computation and memory usage become expensive at long sequence lengths.

This PR adds PyTorch FlexAttention support to VeOmni so these models can use native `BlockMask` sparsity while retaining VeOmni's sequence-parallel attention path.

### What does this PR do?

- Adds a FlexAttention backend with native `BlockMask` and GQA support.
- Adds a unified `fused_attention_forward(...)` facade for FlashAttention and FlexAttention.
- Extracts shared Ulysses attention transport used by both backends.
- Registers the VeOmni Flex attention and mask-builder entries with Transformers.
- Adds Gemma 3 270M text-model support through VeOmni's standard patchgen model path.
- Adds correctness, distributed, fused-loss, and opt-in performance tests.

### FlexAttention and unified `fused_attention_forward`

The existing `flash_attention_forward(...)` remains the FlashAttention adapter and keeps its FA2/FA3/FA4 backend dispatch. A new `flex_attention_forward(...)` adapter consumes a caller-provided native `BlockMask` and dispatches to a replaceable FlexAttention compute backend.

`fused_attention_forward(...)` is the model-facing entry point. It reads `module.config._attn_implementation` and dispatches to the FlashAttention or FlexAttention adapter without recursively querying the Transformers attention registry.

Shared Ulysses sequence/head exchange, GQA KV handling, auxiliary-head slicing, and output restoration live in `veomni/ops/kernels/attention/ulysses.py`. When `MODELING_BACKEND=veomni`, selecting `flex_attention` is rewritten to the SP-aware `veomni_flex_attention_with_sp` implementation.

The VeOmni Flex name is also registered in Transformers' mask registry through an SP-aware wrapper around the native Flex mask builder. Outside Ulysses, it delegates directly to Transformers. Under Ulysses, it builds the native `BlockMask` at global sequence dimensions from the local query length and full-sequence 2D attention mask. The attention kernel validates and consumes that `BlockMask`; `sliding_window` remains model metadata and is not used to reconstruct visibility inside the kernel.

#### Test

Focused Flash/Flex/facade contracts:

```text
30 passed, 3 skipped
```

The tests cover adapter signatures and dispatch, registry/config rewrites, native `BlockMask` validation, GQA, mixed two-dimensional visibility, MATH SDPA output/gradient parity, and Ulysses helper contracts. The real distributed Ulysses suite passes at `4 passed`, including FlashAttention and FlexAttention output and Q/K/V-gradient parity.

Quality and generated-code checks pass:

```bash
PATH="$PWD/.venv/bin:$PATH" make quality
PATH="$PWD/.venv/bin:$PATH" patchgen --check
git diff --check
```

Single-A100 BAGEL-like mixed-visibility results use BF16, batch size 1, forward + backward, and five steady-state iterations. SDPA is forced to `SDPBackend.EFFICIENT_ATTENTION` with no fallback.

| Seq. length | SDPA (EFFICIENT_ATTENTION) | SDPA peak allocated | FlexAttention | Flex peak allocated |
|---:|---:|---:|---:|---:|
| 4,096 | 16.89 ms | 0.499 GiB | 7.39 ms | 0.398 GiB |
| 8,192 | 57.37 ms | 0.996 GiB | 20.66 ms | 0.700 GiB |
| 20,000 | 333.89 ms | 2.955 GiB | 103.41 ms | 1.572 GiB |

The profiling test is opt-in:

```bash
RUN_FLEX_ATTENTION_PROFILE=1 CUDA_VISIBLE_DEVICES=0 \
  .venv/bin/pytest -q tests/ops/test_flex_attention_contract.py \
  -m benchmark -vv -s
```

### Gemma 3 270M text implementation

Gemma 3 is integrated as `gemma3_text` through VeOmni's standard `MODELING_REGISTRY` and patchgen-generated Transformers model path. Its upstream alternating sliding-window/full-attention structure is unchanged and continues to build masks through Transformers.

The model patch only adapts `Gemma3ForCausalLM.forward(...)` to VeOmni's causal-LM loss/output contract, including eager and fused-linear cross entropy. Attention, RMSNorm, MLP, RoPE, softcapping, and full/sliding mask construction remain owned by the Transformers implementation.

A ready-to-run configuration is provided at `configs/text/gemma3.yaml` and uses VeOmni's standard text-training task entry:

```bash
bash train.sh tasks/train_text.py configs/text/gemma3.yaml
```

#### Test

Gemma 3 correctness tests pass at `5 passed` and cover:

- model registry/build and Flex config routing;
- native full/sliding `BlockMask` construction and visibility semantics;
- BF16 MATH SDPA versus VeOmni FlexAttention output, input-gradient, and Q/K/V/O/LM-head gradient parity;
- eager and Liger fused-linear causal-LM loss with finite backward gradients.

A real-checkpoint training smoke through `tasks/train_text.py` also completed successfully on four A100s with FSDP2 and DP2 × SP2. All six optimizer steps produced finite loss and gradient norms, and the run exited cleanly after saving distributed and Hugging Face checkpoints.

Single-A100 Gemma 3 model-core results use BF16, batch size 1, forward + backward, and five steady-state iterations. SDPA is forced to `SDPBackend.EFFICIENT_ATTENTION` with no fallback.

| Seq. length | SDPA (EFFICIENT_ATTENTION) | SDPA peak allocated | FlexAttention | Flex peak allocated |
|---:|---:|---:|---:|---:|
| 4,096 | 271.40 ms | 4.768 GiB | 323.19 ms | 4.090 GiB |
| 8,192 | 792.20 ms | 10.204 GiB | 683.54 ms | 7.907 GiB |
| 20,000 | 3,275.34 ms | 31.447 GiB | 2,703.34 ms | 19.027 GiB |

The profiling test passes at `3 passed` and is opt-in:

```bash
RUN_GEMMA3_FLEX_PROFILE=1 CUDA_VISIBLE_DEVICES=0 \
  .venv/bin/pytest -q tests/models/test_gemma3_flex_attention.py \
  -m benchmark -vv -s
```

### API and Usage Example

For a new model, FlexAttention is not enabled by config alone: its patchgen-generated modeling must dispatch through the Transformers attention/mask registries, construct a model-correct native `BlockMask`, and register the generated class under the config's `model_type`. The configuration below is the final backend-selection step after that model integration; see `docs/transformers_v5/veomni_fused_attention.md` and `docs/design/patchgen.md#adding-a-new-model`.

Select FlexAttention through the existing ops configuration:

```yaml
model:
  ops_implementation:
    attn_implementation: flex_attention
```

With `MODELING_BACKEND=veomni`, this resolves to:

```text
veomni_flex_attention_with_sp
```

The model supplies a native `torch.nn.attention.flex_attention.BlockMask`. When Ulysses is enabled, the BlockMask must currently be head-broadcast (`BlockMask.shape[1] == 1`).

### CODE REVIEW

File: veomni/ops/kernels/attention/ulysses.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    assert query_head_count % ulysses_size == 0, (
+        f"num_query_heads ({query_head_count}) must be divisible by ulysses_size ({ulysses_size})"
+    )
+    if ulysses_size > key_value_head_count:
+        assert ulysses_size % key_value_head_count == 0, (
+            f"ulysses_size ({ulysses_size}) must be divisible by num_key_value_heads ({key_value_head_count})"
+        )
+        repeat_count = ulysses_size // key_value_head_count
+        key = torch.repeat_interleave(key, dim=2, repeats=repeat_count)
+        value = torch.repeat_interleave(value, dim=2, repeats=repeat_count)
```
[the remark this answers is not in the document]
Comment (Zx55) [reply veomni/ops/kernels/attention/ulysses.py:48]: Addressed in `53715bd0`. `prepare_ulysses_qkv` now checks `num_key_value_heads % ulysses_size == 0` when the KV head count is greater than or equal to the Ulysses size, before entering the all-to-all operation. The existing divisibility check and KV-head repetition path for `ulysses_size > num_key_value_heads` remain unchanged.

**Comment by Coach257** [comment]
## Packing × FlexAttention: cross-sample attention leakage

**Severity:** critical (silent training correctness)

VeOmni’s default packing path always supplies a full-length, all-ones `attention_mask` (for FA varlen / `cu_seqlens`). Under Transformers 5.9, FlexAttention only derives a packing-boundary predicate from `position_ids` when `attention_mask is None`:

```python
# transformers/masking_utils.py
if position_ids is not None and attention_mask is None and past_key_values is None:
    packed_sequence_mask = find_packed_sequence_indices(position_ids)
```

So with packing enabled, FlexAttention never gets a document/sample boundary mask. FlashAttention is fine because isolation comes from `cu_seq_lens`; Flex has no equivalent. Tokens from different packed samples can attend to each other.

`configs/text/gemma3.yaml` currently avoids this by setting `dyn_bsz: false` and `micro_batch_size: 1`, but that’s config-side avoidance—not a framework fix. Existing Flex tests don’t cover multi-sample packing.

### Suggested fix (either)

1. **Correct:** when building the Flex `BlockMask` (Gemma / mask path), explicitly AND in  
   `packed_sequence_mask_function(find_packed_sequence_indices(position_ids))`  
   so packing boundaries are enforced even with an all-ones `attention_mask`.

2. **Fence:** if (1) is out of scope for this PR, fail fast in config validation when  
   `attn_implementation` is `flex_attention` / `veomni_flex_attention_with_sp` and packing is active (`dyn_bsz=True` or `micro_batch_size > 1`), and document the limitation in `docs/transformers_v5/veomni_fused_attention.md`.

**Comment by Zx55** [comment]
@Coach257 

Thanks for pointing this out. The fix includes two parts:

1. Added VeOmni mask wrappers in `veomni/models/transformers/masking_utils.py` for HF’s `create_causal_mask` and `create_sliding_window_causal_mask`.

For `veomni_flex_attention_with_sp`, the wrappers convert `cu_seq_lens_q` into token-level packed sequence IDs, combine HF’s `packed_sequence_mask_function` with the existing causal/sliding predicate, and then construct the global `BlockMask`. This prevents attention across packed samples while preserving the original HF behavior for non-Flex backends and non-packed inputs.

2. Updated `gemma3_gpu_patch_gen_config.py` to replace the HF mask imports with the VeOmni wrappers and override `Gemma3TextModel.forward`.

The override forwards `cu_seq_lens_q` into the mask kwargs used by both the full-attention and sliding-window mask builders, so packed boundaries are available when their `BlockMasks` are constructed.

Validation:

- Full and sliding-window BlockMasks both reject cross-sample attention.
- Real CUDA packed [3, 5] forward matches two independent Gemma3 forwards.
- CUDA backward completes with finite input and key parameter gradients.
- Real 2-rank Ulysses/SP forward and backward match the non-SP reference using a global pack-aware BlockMask.
- Related test suite: 47 passed, 6 skipped.

**Review by Coach257 — APPROVED** [review]
